velox/exec/TableScan.cpp

```diff
@@ -436,6 +436,10 @@ void TableScan::addDynamicFilter(
 void TableScan::close() {
   Operator::close();
 
+  if (dataSource_ != nullptr) {
+    dataSource_->close();
```

Comment:
There are also data sources in splits, it's not convenient to close it explicitly, can we just add the cleanup logic in destructor?
https://github.com/facebookincubator/velox/blob/756f104e27d25c3ac29f7f392516f8231f422cff/velox/connectors/Connector.h#L65

Reply:
> can we just add the cleanup logic in destructor?

https://github.com/facebookincubator/velox/pull/12609 is a fix when destructor is relied for the purpose. While the reason it's not that applicable is explained [here](https://github.com/facebookincubator/velox/pull/12609#pullrequestreview-2687358329).

> There are also data sources in splits, it's not convenient to close it explicitly

There is no code in `DataSource::close` yet by default. Do you mean we'd take care of async close as well in the default implementation?

Reply:
@Yuhta do you mean async split source? Each data source is assumed to have one split?

Reply:
Chat with @zhztheplayer offline, can we provide a cancel API to data source for best effort cancellation support and default is do nothing. @zhztheplayer use this in a customized connector for a streaming service.